airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,279 @@
+      # This is an edge case.  In the standard case where there's only a single primary key column, 
+      # this could should only add a single filter
+      for key in self.primary_key_column:
+        last_value_seen = self.primary_key_last_value_seen[key] if key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        if primary_key_filter != "":
+          primary_key_filter = primary_key_filter + " AND "
+        if primary_key_filter == "" and self.has_composite_primary_key():
+          primary_key_filter = primary_key_filter + f"{key} = {last_value_seen}"
+        else:
+          primary_key_filter = primary_key_filter + f"{key} > {last_value_seen}"
+      # if it's a composite key, we add a second column to try to "upgrade" to the next column value
+      # if we don't do this, we don't properly make our way up the table
+      # NOTE:  This only works when there are two columns in the primary key
+      if self.has_composite_primary_key():
+        first_column = self.primary_key_column[0]
+        first_column_value = self.primary_key_last_value_seen[first_column] if first_column in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        primary_key_filter = primary_key_filter + f" OR {first_column} > {first_column_value}"
+      return primary_key_filter
+    
+    def generate_primary_key_sorter(self):
+      primary_key_sorter = ""
+      for key in self.primary_key_column:
+        if primary_key_sorter != "":
+          primary_key_sorter = primary_key_sorter + ", "
+        primary_key_sorter = primary_key_sorter + f"{key} ASC"
+      return primary_key_sorter
+
+    def generate_ordered_query(self, stream_slice):
+      values = ', '.join(self.properties) # we use values instead of '*' because we want to preserve the order of the columns
+      incremental_column_sorter = f", {self.incremental_column} ASC" if self.incremental_column else ""
+      # per https://docs.oracle.com/en/cloud/saas/netsuite/ns-online-help/section_156257805177.html#subsect_156330163819
+      # date literals need to be wrapped in a to_date function
+      incremental_column_filter = f"{self.incremental_column} >= to_timestamp('{stream_slice['first_day']}', 'YYYY-MM-DD') AND {self.incremental_column} <= to_timestamp('{stream_slice['last_day']}', 'YYYY-MM-DD')" if self.incremental_column else ""
+      primary_key_filter = self.generate_primary_key_filter()
+      primary_key_sorter = self.generate_primary_key_sorter()
+      and_connector = " AND " if primary_key_filter != "" and incremental_column_filter != "" else ""
+      where_clause = "WHERE " if primary_key_filter != "" or incremental_column_filter != "" else ""
+      query = f"""
+        SELECT TOP {NETSUITE_PAGINATION_INTERVAL} {values} FROM {self.table_name} 
+        {where_clause}{primary_key_filter}{and_connector}{incremental_column_filter}
+        ORDER BY {primary_key_sorter}""" + incremental_column_sorter
```

Comment:
Incremental syncs are tricky and extremely difficult to debug so I want to make sure we get this right.

Here's my understanding:

Let's assume the page size are 1 record for our own sanity and the table has the following records:

- id: "A", timestamp: 2021-01-02
- id: "B", timestamp: 2021-01-01
- id: "C", timestamp: 2021-01-3

If the sync fails after processing the page, the state emitted will be:
```
{
"primary_key_last_value_seen": "A",
"last_date_updated": 2021-01-02
}
```

Then on a retry, we'll only fetch data with timestamp >= 2021-01-02, and we'll miss out on the record "B".

Is that right? I think this should sort by `timestamp, pk`, but let me know if I missed something